### ISSUE TITLE
Code Lens support

### ISSUE THREAD

**Issue by Trass3r** [clangd/clangd#442]
https://github.com/microsoft/language-server-protocol/blob/master/versions/protocol-2-x.md#textDocument_codeLens

For reference ccls already implements this.

Possibly related to https://github.com/clangd/clangd/issues/162#issuecomment-547515424_

**Comment by Trass3r** [comment]
WIP:
![grafik](https://user-images.githubusercontent.com/404623/87881040-d0c60e80-c9f6-11ea-8e8e-a345842fccea.png)

**Comment by Trass3r** [comment]
@sam-mccall @HighCommander4
It's unclear how different kinds of cross-TU-references like number of base classes and overrides can be distinguished as clangd uses a [simplified](https://github.com/llvm/llvm-project/blob/d1271127240b0920e8758519d95a948d03a832ad/clang-tools-extra/clangd/index/SymbolCollector.cpp#L171) RefKind enum. See this [POC](https://github.com/Trass3r/llvm-project/commit/d86d67b07ecf6fc3afc23196017da7ea008f7614#diff-e968ec3c16e5dcbf4f73863566f1b239L1138).

Also it seems quite wasteful to do a full references lookup just to get a count and then another lookup when the code lens is clicked.

**Comment by kadircet** [comment]
Hi @Trass3r, first of all thanks for giving this a try!

> It's unclear how different kinds of cross-TU references like number of base classes and overrides can be distinguished as clangd uses a simplified RefKind enum. See this POC.

We've got `Relations` in clangd to support such queries rather than references. You might want to take a look at those.

> Also it seems quite wasteful to do a full references lookup just to get a count and then another lookup when the code lens is clicked.

I don't really know about the details of your implementation, nor the design. In theory aren't these two totally independent requests?
What are you planning to return when the codelens is clicked? If truly necessary can't we just attach some `data` to prevent an additional index query?

In addition to all of these, if you are planning to contribute this back to upstream, it would be great to have some sort of a design sketch to understand what's going on in a higher level and what you are trying to achieve.

**Comment by Trass3r** [comment]
> We've got `Relations` in clangd to support such queries rather than references. You might want to take a look at those.

Thanks for the hint.

> I don't really know about the details of your implementation, nor the design. In theory aren't these two totally independent requests?
> What are you planning to return when the codelens is clicked? If truly necessary can't we just attach some `data` to prevent an additional index query?

I just reused the references code I found to test the whole thing. Basically you need a lightweight way to get number of refs for all interesting symbols _in the whole file_ (not just the easily accessible top-level ones I use currently) and when a code lens is clicked you issue a filtered references request.

**Comment by kadircet** [comment]
> Basically you need a lightweight way to get number of refs for all interesting symbols in the whole file

unfortunately, that's not a use case we can support with our current infra easily :( This is another good reason for an sqlite like dbms for the index.

Currently the best you can do is a batched refs request to index and hope that it will be smart about it (currently it isn't maybe in the future). An even bigger improvement would be to somehow support meta queries like refs count and such.

**Comment by Trass3r** [comment]
Ah now that I look at the spec again I think it's not as bad after all: https://microsoft.github.io/language-server-protocol/specification#codeLens_resolve
You just need to return the locations of all interesting declarations in the first step and then it will request the code lenses for the ones currently seen by the user (probably). But still a bit wasteful if none is ever clicked.

**Comment by lightmelodies** [comment]
@Trass3r  Some suggestion for https://github.com/Trass3r/llvm-project/commit/2d95594fe9787212da4b4cbe7254fdf1e650c67b

-  `getLocalTopLevelDecls` does not handle inner declarations. `getDocumentSymbols` will be a good choice.
- `findReferences` does not work for type hierarchy since drived class may not be referenced by base class.  We can use the existed `getTypeHierarchy` function  to collect correct informations . We may also need implement `getMemberHierarchy` / `getCallHierarchy` for method overrides / function calls, just like ccls.

Maybe you will be interested in https://github.com/lightmelodies/llvm-project/commit/6037007e529002eb1b6d1b59c4549e9ed2064024 and  https://github.com/lightmelodies/llvm-project/commit/5c832eba3685d7ec4f95b20583494dbb5a252e22 .

**Comment by Trass3r** [comment]
Yeah I already noticed `getLocalTopLevelDecls` doesn't even cover simple class methods.

> https://microsoft.github.io/language-server-protocol/specification#codeLens_resolve
> You just need to return the locations of all interesting declarations in the first step and then it will request the code lenses for the ones currently seen by the user (probably). But still a bit wasteful if none is ever clicked.

The thing is how do you add multiple code lenses that way?

**Comment by lightmelodies** [comment]
@Trass3r  use `getDocumentSymbols` to get all declarations. For each declaration, return all possible code lenses(the same range, but different commmad). When resolving, use range and command info to get the actual reference count.

**Comment by lightmelodies** [comment]
Almost done.
https://github.com/lightmelodies/llvm-project/commit/c142626d0e32031064b827357b93520612feb97c
https://github.com/lightmelodies/vscode-clangd/commit/c00f74d0d2403a4753dab8b8a0a8dbc8bb6ffdf4
![image](https://user-images.githubusercontent.com/20811066/91650874-c7c96200-eab7-11ea-8aff-8537a1492820.png)

**Comment by Trass3r** [comment]
Nice! But you should have announced you were going to continue work on this. I'm lucky I didn't touch it in the meantime.
Does it make sense though to have a `bases` lens? You see it directly in the code anyway.
And we really need to implement the resolver so those expensive lookups are only done lazily.

**Comment by Trass3r** [comment]
@lightmelodies any updates?

**Comment by Trass3r** [comment]
Just for the record, been using this in production so far without problems.

**Comment by krishnakumarg1984** [comment]
Has there been any news regarding codelens support with `clangd`?

**Comment by Trass3r** [comment]
See my comments. The patch still has some issues that need to be resolved but is stable.
For example
- "base" refs are only useful on class forward decls. PoC:
https://github.com/Trass3r/llvm-project/commit/0a76ee1b597f5c483316de9598bc0f4fc839e13a
- The "refs" include a reference to the declaration itself, so unused functions are confusingly annotated with "1 ref"

But there has been surprising little interest in this major feature so far.
@sam-mccall?

**Comment by HighCommander4** [comment]
@Trass3r if your patch is useful in its current state, I would encourage you to post it to Phabricator for review

**Comment by Trass3r** [comment]
Well as I said it is usable but still has some minor issues. I've been using a patched version of clangd for the last year.
But I don't have the time or maybe even expertise to finish it. So I don't know if it makes sense to push it to Phabricator, if someone will pick it up then.
In any case feel free to comment on Github directly if you notice anything: https://github.com/Trass3r/llvm-project/commit/fa05676f6bdcaa53db8696e18baf7747997ba15e

**Comment by krishnakumarg1984** [comment]
It's surprising that there isn't much demand from the `clangd` developer community to take this forward. `codelens` has been a part of the standard LSP specification for a while now.

**Comment by Trass3r** [comment]
@HighCommander4 looks like there was a revision submitted back then already: https://reviews.llvm.org/D91930
Current state is https://reviews.llvm.org/D131295 as RFC.

**Comment by sparlund** [comment]
@Trass3r is still being worked on? For me personally it's the most important feature that is missing.

**Comment by Trass3r** [comment]
Just minor progress. There were design discussions in D91930 and there's still the issue of multiple code lenses appearing sometimes when templates are involved.

**Comment by sparlund** [comment]
Any progress? @Trass3r

**Comment by Trass3r** [comment]
Nope.
https://github.com/llvm/llvm-project/compare/main...Trass3r:llvm-project:main~40

**Comment by Trass3r** [comment]
https://github.com/llvm/llvm-project/compare/main...Trass3r:llvm-project:codelens
